zingolib/src/wallet/transaction_records_by_id.rs

```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
     pub fn from_map(map: HashMap<TxId, TransactionRecord>) -> Self {
         TransactionRecordsById(map)
     }
 }
+/// This block implements methods to modify the map.
+impl TransactionRecordsById {
+    /// All information after a certain height is invalidated during a reorg.
+    pub fn invalidate_all_transactions_after_or_at_height(&mut self, reorg_height: BlockHeight) {
+        // First, collect txids that need to be removed
+        let txids_to_remove = self
+            .values()
+            .filter_map(|transaction_metadata| {
+                if transaction_metadata
+                    .status
+                    .is_confirmed_after_or_at(&reorg_height)
+                    || transaction_metadata
+                        .status
+                        .is_broadcast_after_or_at(&reorg_height)
+                // tODo: why dont we only remove confirmed transactions. unconfirmed transactions may still be valid in the mempool and may later confirm or expire.
+                {
+                    Some(transaction_metadata.txid)
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+        self.invalidate_txids(txids_to_remove);
+    }
+    /// this function invalidiates a vec of txids by removing them and then all references to them.
+    pub(crate) fn invalidate_txids(&mut self, txids_to_remove: Vec<TxId>) {
+        for txid in &txids_to_remove {
+            self.remove(txid);
+        }
+
+        self.invalidate_txid_specific_transparent_spends(&txids_to_remove);
+        // roll back any sapling spends in each invalidated tx
+        self.invalidate_txid_specific_domain_spends::<SaplingDomain>(&txids_to_remove);
+        // roll back any orchard spends in each invalidated tx
+        self.invalidate_txid_specific_domain_spends::<OrchardDomain>(&txids_to_remove);
+    }
+    /// any transparent note that is listed as being spent in one of these transactions will be marked as unspent.
+    pub(crate) fn invalidate_txid_specific_transparent_spends(
+        &mut self,
+        invalidated_txids: &[TxId],
+    ) {
+        self.values_mut().for_each(|transaction_metadata| {
+            // Update UTXOs to roll back any spent utxos
+            transaction_metadata
+                .transparent_notes
+                .iter_mut()
+                .for_each(|utxo| {
+                    if utxo.is_spent() && invalidated_txids.contains(&utxo.spent().unwrap().0) {
+                        *utxo.spent_mut() = None;
+                    }
+
+                    if utxo.unconfirmed_spent.is_some()
+                        && invalidated_txids.contains(&utxo.unconfirmed_spent.unwrap().0)
+                    {
+                        utxo.unconfirmed_spent = None;
+                    }
+                })
+        });
+    }
+    /// any shielded note of the specified domain that is listed as being spent in one of these transactions will be marked as unspent.
```

Comment:
suggested comment change 
```suggestion
    /// Reverts any spent shielded notes in the given transactions to unspent.
    ///
    /// This is required in the case that the note was spent in an invalidated transaction.
    /// Takes a slice of txids corresponding to the invalidated transactions.
```